Pass vlan tagged packets back to the kernel stack

kvaster opened the pull request:
Pass vlan tagged packets back to the kernel stack.

VLAN tagged packet will be catched with bpf on main interface first. We
need to passs it back to the kernel stack. Tag will be stripped and
packet will be reenqueued on proper interface.

Fixes: #14579

Signed-off-by: Viktor Kuzmin <kvaster@gmail.com>

```release-note
Make NodePort BPF to work on VLAN devices
```

brb commented on bpf/bpf_host.c:
```diff
@@ -916,6 +916,11 @@ handle_netdev(struct __ctx_buff *ctx, const bool from_host)
 	__u16 proto;
 
+	/* Pass vlan tagged packets to the stack - kernel will strip vlan tag and we'll receive it back on proper interface */
+	if (ctx->vlan_present) {
```
It's in the fast path, so let's do the check only if `from_host == false`.

kvaster replied:
You're right. And even more. I think that we need to skip packet processing on egress in to_netdev function. PR update coming up.

brb replied:
> skip packet processing on egress in to_netdev function

What's the packet flow for egress in terms of VLAN devices in the kernel?

kvaster replied:
wan0 - physical interface, wan0.4000 - vlan interface attached to wan0
On ingress packet will appear with vlan tag on wan0 first and then on wan0.4000 without tag.
On egress packet will appear on wan0.4000 first without vlan tag and then on wan0 with tag.

kvaster replied:
In:
wan0:
```
52:54:00:63:7b:cc > 52:54:00:12:34:56, ethertype 802.1Q (0x8100), length 72: vlan 4000, p 0, ethertype IPv4 (0x0800)
10.118.13.1.35330 > 10.118.11.20.5678
```
wan0.4000:
```
52:54:00:63:7b:cc > 52:54:00:12:34:56, ethertype IPv4 (0x0800), length 68
10.118.13.1.35330 > 10.118.11.20.5678
```

Out:
wan.4000:
```
52:54:00:12:34:56 > 52:54:00:12:34:58, ethertype IPv4 (0x0800)
10.118.13.1.35330 > 10.244.2.67.5678
```
wan0:
```
52:54:00:12:34:56 > 52:54:00:12:34:58, ethertype 802.1Q (0x8100), length 72: vlan 4000, p 0, ethertype IPv4 (0x0800)
10.118.13.1.35330 > 10.244.2.67.5678
```

brb replied:
OK, then it makes sense to skip the processing.

brb reviewed (changes requested):
Thanks for the fix! Minor comment. Also, how did you test it?

brb reviewed (approved):
LGTM, thanks!

Let's get @borkmann review too, as he might have more insight into the VLAN handling in the kernel.

borkmann commented on bpf/bpf_host.c:
```diff
@@ -916,6 +916,10 @@ handle_netdev(struct __ctx_buff *ctx, const bool from_host)
 
+	/* Pass vlan tagged packets to the stack - kernel will strip vlan tag and we'll receive it back on proper interface */
+	if (!from_host && ctx->vlan_present)
+		return CTX_ACT_OK;
```
Could you describe some more the expected packet path with this commit for vlan tagged packets? Maybe via ascii drawing. The comment is a little too terse to judge. Also, this breaks BPF host routing, so we'd need to process these differently there since we cannot pass up the stack. How will this work in tunneling mode (e.g. vxlan)?

kvaster replied:
wan0 - original interface wan0.4000 - vlan interface, attached to wan0.

BPF programs attached on both wan0 and wan0.4000 interfaces.
wan0.4000 interface is used as primary (internal) network.
PT - tagged packet, P - normal untagged packet.

Original packet path:
```
* PT -> ingress on wan0
* PT -> cilium BPF ingress
* PT -> assigning pod's destination ip
* PT -> egress on wan0.4000
* PTT -> egress on wan0
```

Cilium makes no difference between tagged and untagged packets, cause the only difference with usual packets are 'vlan_*' fields which are not checked currently. As a result, packet received on wan0 will be load balanced to pod.
Packet will be sent through proper interface - wan0.4000 and kernel will add one more vlan tag for this packet and will pass it to egress on wan0.

Proper path (from my point of view):
```
* PT -> ingress on wan0
* PT -> pass back to kernel
* P -> ingress on wan0.4000
* P -> cilium BPF ingress
* P -> assigning pod's destination ip
* P -> egress on wan0.4000
* PT -> egress on wan0
```

I'm not very good at BPF, but I don't completely understand how it can affect BPF host routing or tunneling mode, cause the only place where vlan tagged packets can appear is physical interface currently. The idea here is: if we receive tagged packet then pass it back to kernel and kernel will do it's job - will strip vlan tag and will start processing packet on proper interface. And if we have vlan tagged packet on egress usually means that we've already processed this packet when it was untagged...

brb replied:
Is it possible to configure the kernel in a way that after `* PT -> pass back to kernel` the packet bypasses bpf_host @ wan0.4000 ingress? If yes, then there is a security concern that the host firewall (implemented in bpf_host) can be bypassed.

kvaster replied:
I think that there is no way for packet to bypass bpf_host@wan0.4000 ingress.

borkmann replied:
> The idea here is: if we receive tagged packet then pass it back to kernel and kernel will do it's job - will strip vlan tag and will start processing packet on proper interface.

But then what happens if the NIC supports offloaded vlan, populates ctx->vlan_* fields (and has the tag already removed from the packet), then pushes up the stack with this PR, but there is no wan0.4000 device (or any other vlan typed netdev, fwiw)?

borkmann replied:
The thing I'm a bit confused about is that almost all modern NICs will populate ctx->vlan_* fields via __vlan_hwaccel_put_tag() and read the vlan info from it's RX descriptor 
 (https://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git/tree/drivers/net/ethernet/intel/ice/ice_txrx_lib.c#n179), so it's not part of the packet data to begin with. What NIC is backing your wan0?

kvaster replied:
On real hardware I have 'Intel Corporation I210 Gigabit'. And I can see same behaviour on my VMs. Also there is no 802.1q header already in skb in handle_netdev function - only vlan_* fields. But I definitely should check what will happen in case there is no any/proper vlan typed device. Packet should be dropped, but I'm not sure.

kvaster replied:
@borkmann , I have an answers for one question:

vlan packet is dropped by kernel always in case there is no proper vlan typed netdev. That means there is no security problem in case we will pass vlan targeted packets back to the kernel stack when we are processing them on main (physical) netdev. Such packets will be processed on proper vlan typed interface after that or will be dropped.

For now I'm investigating how it works with xdp and will get back. Thanks.

kvaster committed:
Pass vlan tagged packets back to the kernel stack.

VLAN packets will be catched with bpf on main interface first. We
need to passs it back to the kernel stack. VLAN info/tag will be stripped and
packet will be reenqueued on proper interface or dropped by kernel.

Also we do not need to process VLAN packets on egress, cause either
we've already processed such packets in case bpf program is attached
to VLAN interface or we do not need to process such packets at all in
other way.

Fixes: #14579

Signed-off-by: Viktor Kuzmin <kvaster@gmail.com>
